File: 132_pattern.py
class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        
        st = []
        cur_min = nums[0]

        for n in nums[1:]:
            while st and n >= st[-1][0]:
                st.pop()
            
            if st and n > st[-1][1]:
                return True
            
            st.append([n, cur_min])
            cur_min = min(cur_min, n)
        
        return False

File: 132_pattern.cpp
class Solution {
public:
    bool find132pattern(vector<int>& nums) {
        stack<pair<int, int>> st;
        int cur_min = nums[0];

        for (size_t i = 1; i < nums.size(); ++i) {
            int n = nums[i];
            while (!st.empty() && n >= st.top().first) {
                st.pop();
            }

            if (!st.empty() && n > st.top().second) {
                return true;
            }

            st.push(std::make_pair(n, cur_min));
            cur_min = std::min(cur_min, n);
        }

        return false;        
    }
};
